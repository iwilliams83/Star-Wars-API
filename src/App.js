import React, { Component } from "react";
import { characters } from "./characters.json";
import Character from "./Character";
import Films from "./Films";

import "./styles.css";

class App extends Component {
  state = {
    selectedCharacter: "",
    filmData: [],
    loaded: false,
    error: false
  };

  getCharacterInfo = (url, name) => {
    this.setState({ loaded: false, selectedCharacter: name });
    fetch(url)
      .then(response => {
        if (response.ok) {
          this.setState({ error: false });
          return response.json();
        }
        this.setState({ error: true });
        throw new Error("Data not found.");
      })
      .then(response => {
        this.getFilmData(response.films);
      })
      .catch(error => {
        console.error(error.message);
      });
  };

  getFilmData = urls => {
    let data = [];
    let apiCalls = [];
    apiCalls = urls.map(url => fetch(url).then(res => res.json()));

    Promise.all(apiCalls).then(response => {
      let options = {
        weekday: "long",
        year: "numeric",
        month: "long",
        day: "numeric"
      };

      data = response.map(res => {
        let date = new Date(res.release_date);
        return {
          title: res.title,
          date: date.toLocaleDateString("en-US", options)
        };
      });
      this.setState({ filmData: data, loaded: true });
    });
  };

  render() {
    const { filmData, loaded, error, selectedCharacter } = this.state;
    return (
      <div className="App">
        <div className="character-container">
          {characters.map(character => {
            return (
              <Character key={character.name}
                character={character}
                getCharacterInfo={this.getCharacterInfo}
              />
            );
          })}
        </div>
        <div>
          {loaded && !error ? (
            <div>
              <Films filmData={filmData} character={selectedCharacter}/>
            </div>
          ) : (
            error && <div className="error-msg"><br/>No data found for {selectedCharacter}</div>
          )}
        </div>
      </div>
    );
  }
}

export default App;
