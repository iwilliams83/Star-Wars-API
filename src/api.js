export const characterAPIcall = url => {
  return fetch(url).then(response => {
    if (response.ok) {
      return response.json()
    }
    throw new Error("Data not found.")
  })
}

export const filmDataAPIcall = urls => {
  let apiCalls = []
  apiCalls = urls.map(url => fetch(url).then(res => res.json()))

  return Promise.all(apiCalls)
}
