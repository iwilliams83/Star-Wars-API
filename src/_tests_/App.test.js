import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import App from "../App";
import * as api from "../api";

jest.mock("../api");

afterAll(() => {
  jest.restoreAllMocks();
});

describe("initial render <App />", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("matches snapshot", () => {
    // https://jestjs.io/docs/en/tutorial-react#snapshot-testing
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("gets character's films", () => {
  it("has results for valid api call", async () => {
    // Arrange
    const characterName = 'Luke Skywalker'
    const characterUrl = 'https://character-api'
    const characterApiMockResponse = { films: ["https://fakeurl1", "https://fakeurl2"]};
    api.characterAPIcall.mockResolvedValue(characterApiMockResponse);

    const filmDataAPIMockResponse = [
      { title: 'the first movie', release_date: '1977-05-25'},
      { title: 'the second movie', release_date: '1979-05-25'},
      { title: 'the third movie', release_date: '1980-05-25'}
    ];
    api.filmDataAPIcall.mockResolvedValue(filmDataAPIMockResponse);

    const options = {
      weekday: "long",
      year: "numeric",
      month: "long",
      day: "numeric"
    };

    const formattedResponse = filmDataAPIMockResponse.map(res => {
      let date = new Date(res.release_date);
      return {
        title: res.title,
        date: date.toLocaleDateString("en-US", options)
      };
    });

    // Act
    const wrapper = shallow(<App />);
    await wrapper
        .find("Character")
        .first()
        .prop('getCharacterInfo')(characterUrl, characterName)

    // Wait for state to update
    await wrapper.update();

    // Assert
    expect(api.characterAPIcall).toBeCalledWith(characterUrl)
    expect(api.filmDataAPIcall).toBeCalledWith(characterApiMockResponse.films)
    expect(wrapper.find('Films').prop('filmData')).toEqual(formattedResponse)
    expect(wrapper.find('Films').prop('character')).toEqual(characterName)
  })

  it("displays error for invalid api call", async () => {
    const characterUrl = 'https://unknown-character'
    const characterName = 'Obi-wan Kenobi'
    const mockError = new Error();
    api.characterAPIcall.mockRejectedValue(mockError);

    const wrapper = shallow(<App />);
    await wrapper
        .find("Character")
        .at(2)
        .prop('getCharacterInfo')(characterUrl, characterName)

    await wrapper.update();

    expect(api.characterAPIcall).toBeCalledWith(characterUrl)
    expect(wrapper.find('.error-msg').text()).toEqual(`No data found for ${characterName}`)
  })
})
