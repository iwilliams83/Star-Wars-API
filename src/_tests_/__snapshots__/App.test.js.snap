// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial render <App /> matches snapshot 1`] = `
<div
  className="App"
>
  <h2
    className="prompt"
  >
    Click on a character to see their film info
  </h2>
  <div
    className="character-container"
  >
    <div
      className="character-card"
      onClick={[Function]}
    >
      <img
        alt="Responsive image"
        className="img-fluid"
        src="luke-skywalker.jpg"
      />
      <h4
        className="character-name"
      >
        Luke Skywalker
      </h4>
    </div>
    <div
      className="character-card"
      onClick={[Function]}
    >
      <img
        alt="Responsive image"
        className="img-fluid"
        src="darth-vader.jpg"
      />
      <h4
        className="character-name"
      >
        Darth Vader
      </h4>
    </div>
    <div
      className="character-card"
      onClick={[Function]}
    >
      <img
        alt="Responsive image"
        className="img-fluid"
        src="obi-wan-kenobi.jpg"
      />
      <h4
        className="character-name"
      >
        Obi-wan Kenobi
      </h4>
    </div>
    <div
      className="character-card"
      onClick={[Function]}
    >
      <img
        alt="Responsive image"
        className="img-fluid"
        src="r2-d2.jpg"
      />
      <h4
        className="character-name"
      >
        R2-D2
      </h4>
    </div>
  </div>
  <div />
</div>
`;
